File: domain/src/main/java/com/khalturin/dmitriy/domain/iteractor/UseCase.java
package com.khalturin.dmitriy.domain.iteractor;

import com.khalturin.dmitriy.domain.executor.PostExecutionThread;
import com.khalturin.dmitriy.domain.executor.ThreadExecutor;

import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Dmitriy Khalturin <dmitry.halturin.86@gmail.com>
 * for android-rss-reader on 26.04.18 22:12.
 */

public abstract class UseCase<T, Params> {

  private final ThreadExecutor mThreadExecutor;
  private final PostExecutionThread mPostExecutionThread;
  private final CompositeDisposable mDisposables;

  public UseCase(ThreadExecutor threadExecutor, PostExecutionThread postExecutionThread){
    this.mThreadExecutor = threadExecutor;
    this.mPostExecutionThread = postExecutionThread;
    this.mDisposables = new CompositeDisposable();
  }

  abstract Observable<T> buildUseCaseObservable(Params params);

  public void execute(DisposableObserver<T> observer, Params params){
    if(observer != null){
      Observable<T> observable = buildUseCaseObservable(params)
        .subscribeOn(Schedulers.from(mThreadExecutor))
        .observeOn(mPostExecutionThread.getScheduler());

      addDisposable(
        observable.subscribeWith(observer)
      );
    }
  }

  public Observable<T> execute(Params params){
    return buildUseCaseObservable(params)
      .subscribeOn(Schedulers.from(mThreadExecutor))
      .observeOn(mPostExecutionThread.getScheduler());
  }

  public void dispose(){
    if(!mDisposables.isDisposed()){
      mDisposables.dispose();
    }
  }

  private void addDisposable(Disposable disposable){
    if(disposable != null){
      mDisposables.add(disposable);
    }
  }

}

File: domain/src/main/kotlin/com/khalturin/dmitriy/domain/interactor/UseCase.kt
package com.khalturin.dmitriy.domain.interactor

import com.khalturin.dmitriy.domain.executor.PostExecutionThread
import com.khalturin.dmitriy.domain.executor.ThreadExecutor
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers

/**
 * Created by Dmitriy Khalturin <dmitry.halturin.86@gmail.com>
 * for android-rss-reader on 24.10.18 21:19.
 */
abstract class UseCase<T, Params>(
  private val mThreadExecutor: ThreadExecutor,
  private val mPostExecutionThread: PostExecutionThread
) {

  abstract fun buildUseCaseObservable(params: Params): Observable<T>

  fun execute(params: Params) = buildUseCaseObservable(params)
    .subscribeOn(Schedulers.from(mThreadExecutor))
    .observeOn(mPostExecutionThread.getScheduler())

  private val mDisposables = CompositeDisposable()

  fun execute(observer: DisposableObserver<T>, params: Params) {
    mDisposables.add(
      execute(params).subscribeWith(observer)
    )
  }

  fun dispose() {
    when (mDisposables.isDisposed) {
      false -> mDisposables.dispose()
    }
  }

}
